filters/diag/histogram.go

```diff
@@ -0,0 +1,123 @@
+// It expects a list of number and duration string pairs that defines a histogram.
+// Duration strings define consecutive bucket boundaries starting from implicit zero and numbers define bucket weights.
+// The filter randomly selects a bucket according to its weight and then sleeps for a random duration in between bucket boundaries.
+// The probability of selecting a bucket equals to its weight divided by the sum of all bucket weights.
+// Eskip example:
+//
+//	r: * -> histogramRequestLatency(50, "5ms", 0, "10ms", 30, "15ms", 20, "20ms") -> "https://www.example.org";
+//
+// The example above adds a latency
+// * between implicit 0ms and 5ms to 50% of the requests
+// * between 5ms and 10ms to 0% of the requests
+// * between 10ms and 15ms to 30% of the requests
+// * and between 15ms and 20ms to 20% of the requests.
+func NewHistogramRequestLatency() filters.Spec {
+	return &histSpec{typ: filters.HistogramRequestLatencyName}
+}
+
+// NewHistogramResponseLatency creates filters that add latency to responses according to the histogram distribution, similar to NewHistogramRequestLatency.
+func NewHistogramResponseLatency() filters.Spec {
+	return &histSpec{typ: filters.HistogramResponseLatencyName}
+}
+
+func (s *histSpec) Name() string {
+	return s.typ
+}
+
+func (s *histSpec) CreateFilter(args []any) (filters.Filter, error) {
+	if len(args) == 0 || len(args)%2 != 0 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	f := &histFilter{
+		response: s.typ == filters.HistogramResponseLatencyName,
+		sleep:    time.Sleep,
+		weights:  make([]float64, len(args)/2),
+		buckets:  make([]time.Duration, len(args)/2),
+	}
+
+	for i := 0; i < len(args); i += 2 {
+		weight, ok := args[i].(float64)
+		if !ok || weight < 0 {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		f.weights[i/2] = weight
+		f.sum += weight
+
+		ds, ok := args[i+1].(string)
+		if !ok {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		max, err := time.ParseDuration(ds)
+		if err != nil || max <= 0 {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		if i > 0 && max <= f.buckets[i/2-1] {
+			return nil, filters.ErrInvalidFilterParameters
+		}
+		f.buckets[i/2] = max
+	}
+
+	if f.sum > 0 {
+		for i := range f.weights {
+			f.weights[i] /= f.sum
+		}
+	}
+
+	return f, nil
+}
+
+func (f *histFilter) Request(filters.FilterContext) {
+	if !f.response && f.sum > 0 {
+		f.sleep(f.sample())
+	}
+}
+
+func (f *histFilter) Response(filters.FilterContext) {
+	if f.response && f.sum > 0 {
+		f.sleep(f.sample())
+	}
+}
+
+func (f *histFilter) sample() time.Duration {
+	r := rand.Float64() // #nosec
+	i := 0
+	var upto, c float64
+	for i, c = range f.weights {
+		upto += c
+		if upto >= r {
+			break
+		}
+	}
+	max := f.buckets[i]
+	var min time.Duration
+	if i > 0 {
+		min = f.buckets[i-1]
+	}
+	return min + time.Duration(rand.Int63n(int64(max-min))) // #nosec
```

Comment:
How does `min + x` match the example:
```
// Eskip example:
//
//	r: * -> histogramRequestLatency(50, "5ms", 0, "10ms", 30, "15ms", 20, "20ms") -> "https://www.example.org";
//
// The example above adds a latency
// * between implicit 0ms and 5ms to 50% of the requests   <--
```

Reply:
Previous loop selects `i`-th bucket according to its weight. This bucket left boundary `min` is `bucket[i-1]` (or zero) and right boundary `max` is `bucket[i]`. Then filter selects a random value from `[min, max)` interval as `rand(min, max) = min + rand(0, max-min)`

Reply:
Where is the implicit 0ms?

Reply:
The example is not wrong, please see docs for explanation:
```
The example above adds a latency
* between implicit 0ms and 5ms to 50% of the requests
* between 5ms and 10ms to 0% of the requests
* between 10ms and 15ms to 30% of the requests
* and between 15ms and 20ms to 20% of the requests.
```

Reply:
If it makes it easier to understand we may change filter to always require `"0ms"` (or any zero duration e.g. "0h") as a first argument.

Reply:
Or I we can actually swap order of weights and duration and require first and last arguments to be durations, e.g:
`histogramRequestLatency("5ms", 10, "10ms", 0, "20ms", 90, "30ms")` would add 5-10ms latency to 10% of requests and 20-30ms to 90%

Reply:
Order of args does not matter.

I think what needs to be clear is min. Either min is the upper bound of the leftmost bucket or the example or code is wrong.
Maybe min is not the right name for it. That's just what is not easy to understand and maybe you should define a min as absolute min?!

Reply:
Thank you.

> I think your last comment would make most sense as user input.

I've updated the code accordingly.